app/src/test/java/com/example/main_animal_villa/CharacterTests.kt

```diff
@@ -0,0 +1,108 @@
+import kotlinx.coroutines.test.setMain
+import org.junit.*
+import org.junit.rules.TestRule
+import java.util.*
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+class CharacterTests {
+
+    @get:Rule
+    var rule : TestRule = InstantTaskExecutorRule()
+    var character : Character? = Character()
+
+    lateinit var characterService : CharacterService
+
+    lateinit var mvm : MainViewModel
+
+    @MockK
+    lateinit var mockCharacterService : CharacterService
+
+    private val mainThreadSurrogate = newSingleThreadContext("Main Thread")
+
+    @Before
+    fun initMocksAndMainThread()    {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(mainThreadSurrogate)
+    }
+
+    @After
+    fun tearDown()  {
+        Dispatchers.resetMain()
+        mainThreadSurrogate.close()
+    }
+
+    @Test
+    fun `given character data is available when I name my character Billy and give them the janitor occupation then Character class should have Billy as name and Janitor as occupation` () = runTest {
+        givenCharacterServiceIsInitialized()
+        whenCharacterDataAreReadAndParsed()
+        thenTheCharacterShouldHaveTheNameBillyAndTheOccupationJanitor()
+    }
+
+    @Test
+    private fun givenCharacterServiceIsInitialized()    {
+        characterService = CharacterService()
+    }
+
+    @Test
+    private suspend fun whenCharacterDataAreReadAndParsed() {
+        character = characterService.fetchCharacter()
+    }
+
+    @Test
+    private fun thenTheCharacterShouldHaveTheNameBillyAndTheOccupationJanitor()    {
+        assertNotNull(character)
+        assertTrue(character.toString().equals("Billy Janitor"))
+    }
+
+    @Test
+    fun `given a view model with live data when populated with a character then result shows character data` () = runTest {
+        givenViewModelIsInitializedWithMockData()
+        whenJSONDataAreReadAndParsed()
+        thenResultsShouldContainCharacter()
+    }
+
+    private fun givenViewModelIsInitializedWithMockData() {
+        val character = Character("Billy", "Janitor")
+
+        coEvery { mockCharacterService.fetchCharacter() } returns character
+
+        mvm = MainViewModel()
+        mvm.characterService = mockCharacterService
+    }
+
+    private fun whenCharacterServiceFetchCharacterInvoked() {
+        mvm.fetchCharacter()
+    }
+
+    private fun thenResultsShouldContainCharacter() {
+        var resultCharacter : Character? = Character()
+        val latch = CountDownLatch(1)
+        val observer = object : Observer<Character> {
+            override fun onChanged(t: Character?) {
+                resultCharacter = t
+                latch.countDown()
+                mvm.character.removeObserver(this)
+            }
+        }
+        mvm.character.ObserveForever(observer)
+        latch.await(10, TimeUnit.SECONDS)
+        assertNotNull(character)
+        assertTrue(character.toString().equals("Billy Janitor"))
+    }
```

Comment:
You can use things such as CountDownLatch to ensure you are getting data from your service or viewmodel. If you 
are retrieving data from a JSON source, you will want to make sure that the function has enough time to receive this data before timing out. CountDownLatch lets you set a timer for however much time you expect the function would need to run consistently.
https://developer.android.com/reference/kotlin/java/util/concurrent/CountDownLatch